iree/tools/utils/cpu_features.c

```diff
@@ -0,0 +1,166 @@
+      : /* clobbers */);                                           \
+    return retval;                                                 \
+  }
+// clang-format on
+
+// Define the access functions for the cpu feature registers that we need.
+IREE_CPU_FEATURES_AARCH64_READ_REGISTER(ID_AA64ISAR0_EL1)
+IREE_CPU_FEATURES_AARCH64_READ_REGISTER(ID_AA64ISAR1_EL1)
+
+// Helper for iree_cpu_features_aarch64_get_register:
+// Immediately reads the value of a cpu feature register, by enum index.
+// Expensive!
+static uint64_t iree_cpu_features_aarch64_read_register(
+    iree_cpu_features_aarch64_register_t idx) {
+  switch (idx) {
+    case iree_cpu_features_aarch64_register_ID_AA64ISAR0_EL1:
+      return iree_cpu_features_aarch64_read_ID_AA64ISAR0_EL1();
+    case iree_cpu_features_aarch64_register_ID_AA64ISAR1_EL1:
+      return iree_cpu_features_aarch64_read_ID_AA64ISAR1_EL1();
+    case iree_cpu_features_aarch64_register_count:  // case, not default, so
+      break;  // the compiler can warn if we forget an enum value.
+  }
+  assert(false && "bad iree_cpu_features_aarch64_register_t value");
+  return 0;
+}
+
+// Returns the value of a cpu feature register. Caches the value in the
+// iree_cpu_features_t struct.
+static uint64_t iree_cpu_features_aarch64_get_register(
+    iree_cpu_features_t* features, iree_cpu_features_aarch64_register_t idx) {
+  if (!features->registers_initialized[idx]) {
+    if (iree_devices_features_aarch64_register_access()) {
+      features->registers[idx] = iree_cpu_features_aarch64_read_register(idx);
+    } /* else: leave registers[idx] with its initial 0 bits */
+    features->registers_initialized[idx] = true;
+  }
+  return features->registers[idx];
+}
+
+// Returns true if +dotprod is supported.
+static bool iree_cpu_features_aarch64_dotprod(iree_cpu_features_t* features) {
+  uint64_t register_value = iree_cpu_features_aarch64_get_register(
+      features, iree_cpu_features_aarch64_register_ID_AA64ISAR0_EL1);
+  return (register_value >> 44) & 1;
+}
+
+// Returns true if +i8mm is supported.
+static bool iree_cpu_features_aarch64_i8mm(iree_cpu_features_t* features) {
+  uint64_t register_value = iree_cpu_features_aarch64_get_register(
+      features, iree_cpu_features_aarch64_register_ID_AA64ISAR1_EL1);
+  return (register_value >> 52) & 1;
+}
+
+typedef iree_cpu_features_t iree_cpu_features_t;
+
+#else  // not defined __aarch64__
+
+// Not-implemented case. Decided in PR #8316 to make it non-empty just to avoid
+// edge cases with empty structs.
+struct iree_cpu_features_t {
+  int unused;
+};
+
+#endif  // defined __aarch64__
+
+iree_status_t iree_cpu_features_create(iree_allocator_t allocator,
+                                       iree_cpu_features_t** cpu_features) {
+  *cpu_features = NULL;
+  IREE_RETURN_IF_ERROR(iree_allocator_malloc(
+      allocator, sizeof(iree_cpu_features_t), (void**)cpu_features));
+  memset(*cpu_features, 0, sizeof(iree_cpu_features_t));
```

Comment:
I believe this `memset` is redundant, since you are not using `iree_allocator_malloc_uninitialized`:

https://github.com/google/iree/blob/7b7e32f7a4167429b350d6d6ac818ddbff563742/iree/base/allocator.h#L233-L243

Reply:
great, thanks!